stdlib/messaging/kafka/src/main/ballerina/src/kafka/producer.bal

```diff
@@ -174,11 +199,176 @@ public type Producer client object {
     # + partition - Partition to which the record should be sent.
     # + timestamp - Timestamp of the record, in milliseconds since epoch.
     # + return - Returns `kafka:ProducerError` if send action fails to send data, nil otherwise.
-    public remote function send(byte[] value, string topic, public byte[]? key = (), public int? partition = (),
+    public remote function send(Data value, string topic, public Data? key = (), public int? partition = (),
                                 public int? timestamp = ()) returns ProducerError? {
-        return producerSend(self, value, java:fromString(topic), key, partition, timestamp);
-    }
 
+        handle topicHandle = java:fromString(topic);
+
+        // Handle string values
+        if (self.valueSerializer == SER_STRING) {
+            if (value is string) {
+                handle valueHandle = java:fromString(value);
+
+                if (key is ()) {
+                    return producerSendString(self, valueHandle, topicHandle, partition, timestamp);
+                }
+
+                if (self.keySerializer == SER_STRING) {
+                    if (key is string) {
+                        handle keyHandle = java:fromString(key);
+                        return producerSendStringString(self, valueHandle, topicHandle, keyHandle, partition,
+                                                        timestamp);
+                    }
+                    panic getKeyTypeMismatchError("string");
+                }
+
+                if (self.keySerializer == SER_INT) {
+                    if (key is int) {
+                        return producerSendStringInt(self, valueHandle, topicHandle, key, partition, timestamp);
+                    }
+                    panic getKeyTypeMismatchError("int");
+                }
+
+                if (self.keySerializer == SER_FLOAT) {
+                    if (key is float) {
+                        return producerSendStringFloat(self, valueHandle, topicHandle, key, partition, timestamp);
+                    }
+                    panic getKeyTypeMismatchError("float");
+                }
+
+                if (self.keySerializer == SER_BYTE_ARRAY) {
+                    if (key is byte[]) {
+                        return producerSendStringByteArray(self, valueHandle, topicHandle, key, partition, timestamp);
+                    }
+                    panic getKeyTypeMismatchError("byte[]");
+                }
+            }
+
+            panic getValueTypeMismatchError("string");
+        }
+
+        // Handle int values
+        if (self.valueSerializer == SER_INT) {
```

Comment:
Shall we move these sub logic to a sub function?

Reply:
Done